File: src/lib/api/manualTests.ts
// Manual Tests (test build only). Every command but `manual_tests_available` exists only in a
// binary built with the `test-build` Cargo feature.
import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export type ManualTestStatus = "pass" | "fail" | "info";

export interface ManualTest {
  id: string;
  title: string;
  description: string;
  changes: string;
  changes_system: boolean;
  /** Default duration in minutes, for a test that runs over time. */
  minutes: number | null;
}

export interface ManualTestReport {
  test_id: string;
  status: ManualTestStatus;
  summary: string;
  details: string[];
  report: string;
}

export interface ManualTestLogEvent {
  test_id: string;
  line: string;
}

export async function manualTestsAvailable(): Promise<boolean> {
  return await invoke<boolean>("manual_tests_available");
}

export async function listManualTests(): Promise<ManualTest[]> {
  return await invoke<ManualTest[]>("list_manual_tests");
}

export async function runManualTest(testId: string, minutes: number | null): Promise<ManualTestReport> {
  return await invoke<ManualTestReport>("run_manual_test", { testId, minutes });
}

export async function cancelManualTest(): Promise<void> {
  await invoke("cancel_manual_test");
}

export async function onManualTestLog(handler: (event: ManualTestLogEvent) => void): Promise<UnlistenFn> {
  return await listen<ManualTestLogEvent>("manual-test-log", (event) => handler(event.payload));
}

File: src-tauri/src/commands/manual_tests.rs
use crate::error::Result;

/// Always registered: the frontend shows the Manual Tests view only when this is `true`.
#[tauri::command]
pub async fn manual_tests_available() -> Result<bool> {
    log::info!("manual_tests_available");
    Ok(cfg!(feature = "test-build"))
}

#[cfg(feature = "test-build")]
#[tauri::command]
pub async fn list_manual_tests() -> Result<Vec<crate::manual_tests::ManualTestView>> {
    log::info!("list_manual_tests");
    Ok(crate::manual_tests::list())
}

#[cfg(feature = "test-build")]
#[tauri::command]
pub async fn run_manual_test(
    app: tauri::AppHandle,
    test_id: String,
    minutes: Option<u32>,
) -> Result<crate::manual_tests::ManualTestReport> {
    log::info!("run_manual_test: '{test_id}' minutes {minutes:?}");
    crate::manual_tests::run(app, test_id, minutes).await
}

#[cfg(feature = "test-build")]
#[tauri::command]
pub async fn cancel_manual_test() -> Result<()> {
    log::info!("cancel_manual_test");
    crate::manual_tests::cancel();
    Ok(())
}
